### ISSUE TITLE
Possibly obvious question about concurrency of the server

### ISSUE THREAD

**Issue by lambrospetrou** [eawsy/aws-lambda-go-net#2]
Hello, forgive me if my question's answer is obvious but it seems I am missing something.

I saw that the listener's Accept() method is protected by a channel (inc). This implies that only one request can be handled each time, hence one Lambda invocation right? 
Why do you have this limitation? Can't we have the net package declare new variables for each request instead of resetting the global vars?

Thanks

**Comment by cristim** [comment]
That's just how Lambda works internally.

Lambda executes the code in a container that is started fresh or resumed for each and every invocation and then paused once it is done, but a single function call is running in the container at a given time.

When the function is called at a high enough rate, the runtime instantiates more such containers, and schedules the runs there in order to handle the extra load. The idle containers get destroyed after a period of inactivity.

See the [documentation](http://docs.aws.amazon.com/lambda/latest/dg/lambda-introduction.html) for more details.

**Comment by lambrospetrou** [comment]
@cristim  You are right! I was confused with something else to be honest on Lambda. Cool, then this package is exactly what I was looking for;p Thanks

ps. Although I would still make the vars private/local instead of global to keep the stateless feature clear;p

**Comment by fsenart** [comment]
@lambrospetrou there are two aspects when running in the AWS Lambda environment:
- Executions are isolated: one and only one request is handled at any time.
- Executions *should* be stateless: goroutines *must* complete before the code exits.

I admit that these aspects are weird to deal with. Mostly the second one as we have noticed very strange behaviors when goroutines does not complete before exit.

That being said, this project aims to provide a `net` listener for AWS Lambda. As such, it should behave like a net interface which respects the two above aspects. So we have a blocking channel to start the processing and a blocking channel to ensure that we won't exit before an answer is returned from the user handler (as Go `http` package internally uses goroutines).

For the other part of your question concerning the private/local vars instead of global vars, this is more a matter of design. In usual cases, you would probably never code something like this, but in AWS Lambda it is more *acceptable*. 
To comply with above two aspects, one need to have some kind of "coordinator". In the case of AWS Lambda global state (along with this bunch of global variables) does the job, *simply*.

PS: At the time being, the project still uses the first version of our Go shim https://github.com/eawsy/aws-lambda-go. But we've recently released a new version which leverages Go 1.8 plugins https://github.com/eawsy/aws-lambda-go-shim. Also, in the next coming days, we will migrate this project to use the new version of the shim. The code should become more simple and clear, and we will try to take into account your remarks. Keep connected ;)
